### PR TITLE
(CDAP-1917) Integrate plugin support to adapter

### PR DESCRIPTION
**Pull request by chtyim**
Public API changes:

- Modifed AdapterConfigurer to have methods for declaring plugins usage.
- Introduce a PluginSelector class for selecting plugin at configure adapter time
- Added AdapterContext to expose access to plugin classes at runtime.
- Have MapReduceContext and WorkerContext to implements AdapterContext for exposing plugin access

Implementation details:

- Have AbstractContext implements AdapterContext.
  - Children classes can provide Adapter spec and PluginInstantiator to enable plugin support
- Modify AbstractDistributedProgramRunner to localize plugin jars to twill container
- Modify AbstractProgramTwillRunnable to setup plugin path in cConf
- Update WorkerProgramRunner to create WorkerContext with plugin support
- Update MapReduceProgramRunner to enable plugin support
  - Involves updating multiple MapReduce context related classes
  - Have the plugin jars send to MR DistributedCache so that all task containers have access to plugins

### CODE REVIEW

**Comment by chtyim** [comment]
unit-test are yet to be added. Will come in next PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java [hunk 1]
```diff
@@ -51,5 +51,7 @@
 
   public static final String ADAPTER_NAME = "adapterName";
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: do we need both ``ADAPTER_NAME`` and ``ADAPTER_SPEC``?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/ProgramOptionConstants.java]: I keep it for now because the `ProgramLifecycleService.start` method is using it for updating `RunRecord` in store. Don't want to introduce too big of a refactoring around the lifecycle service in this PR.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicMapReduceContext.java [hunk 2]
```diff
@@ -72,13 +74,16 @@ public DynamicMapReduceContext(Program program,
...
                                  Arguments runtimeArguments,
                                  MapReduceSpecification spec,
-                                 long logicalStartTime, String workflowBatch, String adapterName,
+                                 long logicalStartTime, String workflowBatch,
                                  DiscoveryServiceClient discoveryServiceClient,
                                  MetricsCollectionService metricsCollectionService,
                                  TransactionSystemClient txClient,
-                                 DatasetFramework dsFramework) {
+                                 DatasetFramework dsFramework,
+                                 @Nullable AdapterSpecification adapterSpec,
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicMapReduceContext.java]: nit: Looks like ``type`` and ``taskId`` are also ``@Nullable``. Unrelated, but if it doesn't increase the size of this PR, maybe we can mark them so as well?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 3]
```diff
@@ -818,13 +847,80 @@ private Location copyProgramJar() throws IOException {
...
+   */
+  @Nullable
+  private Location createPluginArchive(@Nullable AdapterSpecification adapterSpec,
+                                       Id.Program programId) throws IOException {
+    if (adapterSpec == null) {
+      return null;
+    }
+
+    Set<PluginInfo> pluginInfos = adapterSpec.getPluginInfos();
+    if (pluginInfos.isEmpty()) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: When does this happen - I mean that an adapter does not contain any ``pluginInfos``?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: It's not required for adapter to have plugin. Someone can write an app template that doesn't use plugin at all.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java [hunk 4]
```diff
@@ -818,13 +847,80 @@ private Location copyProgramJar() throws IOException {
...
+    }
+
+    File templateDir = new File(cConf.get(Constants.AppFabric.APP_TEMPLATE_DIR));
+    File pluginDir = new File(templateDir, adapterSpec.getTemplate());
+    File jarFile = File.createTempFile("plugin", ".jar");
+    try {
+      JarOutputStream output = new JarOutputStream(new FileOutputStream(jarFile));
+      try {
+        // Create the directory entries
+        output.putNextEntry(new JarEntry(adapterSpec.getTemplate() + "/"));
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Just checked that ``JarEntry`` doesn't take a ``File`` or something like that as input :(

Will this break on windows?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: Jar Entry is always "/" separated. It's abstract path.

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: See http://docs.oracle.com/javase/6/docs/api/java/util/zip/ZipEntry.html#isDirectory()

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java [hunk 5]
```diff
@@ -164,6 +169,50 @@ public TwillController launch(TwillApplication twillApplication) {
   }
 
   /**
+   * Gets plugin files that needs to be localized for the adapter. If the given run is not an adapter, no
+   * modification will be done to the map.
+   */
+  private Map<String, File> addAdapterPluginFiles(ProgramOptions options, Map<String, File> localizeFiles) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: Perhaps call this ``filesToLocalize`` to make it a noun?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: Why needs to be a noun? This method is to add files that needs to be localized to the given map. Shouldn't it call "addXXX" ?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java [hunk 6]
```diff
@@ -110,6 +120,23 @@ public ProgramController run(Program program, ProgramOptions options) {
     return controller;
   }
 
+  @Nullable
+  private AdapterSpecification getAdapterSpecification(Arguments arguments) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java]: Thinking out loud - These methods would potentially be required in every program that is supported through app templates, right? with very similar logic? Is there a way to abstract them?

Perhaps static methods ``AdapterSpecification.from(Arguments)`` (or better ``AdapterSpecification.from(@Nullable String json)``, ``PluginInstantiator.from(@Nullable AdapterSpecification, ClassLoader)``?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java (hunk 5 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: oh... I meant the map, not the method

File: cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java [hunk 7]
```diff
@@ -112,7 +115,64 @@
...
+   *
+   * @param pluginType plugin type name
+   * @param pluginName plugin name
+   * @param properties properties for the plugin. The same set of properties will be used to instantiate the plugin
+   *                   instance at execution time
+   * @param <T> type of the plugin class
+   * @return A new instance of the plugin class or {@code null} if no plugin was found
+   */
+  @Nullable
+  <T> T usePlugin(String pluginType, String pluginName, PluginProperties properties);
```
Comment (gokulavasan) [inline cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java]: Will this be exposed to plugins through PipelineConfigurer for them to depend on other plugins? For example, DBSource plugin might want to add the driver class that it depends on during the configurePipeline stage?

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+  /**
+   * Returns the set of properties available for the plugin when the adapter was created.
+   */
+  public PluginProperties getProperties() {
+    return properties;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java]: can we use Guice here since this is not in API?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java (hunk 6 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java]: 1. No static method (mainly is not good, and partly because the AdapterSpecification is going to move into cdap-api).
2. The current ProgramRunner hierarachy doesn't have an abstract base class. When we need it in many program runner, we can abstract the program runner logic into some abstract base. I can put a TODO there (just don't wanna do it in the PR and don't have time to create a separate PR to refactor it properly).

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java [hunk 9]
```diff
@@ -122,6 +147,101 @@ public void createDataset(String datasetInstanceName,
...
+  public <T> Class<T> usePluginClass(String pluginType, String pluginName, PluginProperties properties) {
+    return usePluginClass(pluginType, pluginName, properties, new PluginSelector());
+  }
+
+  @Override
+  public <T> Class<T> usePluginClass(String pluginType, String pluginName,
+                                     PluginProperties properties, PluginSelector selector) {
+
+    Preconditions.checkArgument(hasPlugin(pluginType, pluginName),
+                                "Plugin of type %s, name %s was already added.", pluginType, pluginName);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java]: This message is misleading? Should it be "...should have been already added" or "...does not exist"?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java (hunk 6 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/worker/WorkerProgramRunner.java]: :+1:

File: cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java (hunk 7 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java]: PipelineConfigurer is internal and specific for ETL app. The CDAP system doesn't know about it. It's the job of the ETL program to expose it through the `PipelineConfigurer` if needed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java (hunk 8 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java]: This class is used inside `AdapterSpecification`, which we are planning to move it to api so that app can access it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java [hunk 10]
```diff
@@ -77,10 +103,17 @@ public AbstractContext(Program program, RunId runId,
...
     // todo: initialized datasets should be managed by DatasetContext (ie. DatasetInstantiator): refactor further
     this.datasets = Datasets.createDatasets(dsInstantiator, datasets, runtimeArguments);
+    this.adapterSpec = adapterSpec;
+    this.pluginInstantiator = pluginInstantiator;
   }
 
   public abstract Metrics getMetrics();
 
+  @Nullable
+  public AdapterSpecification getAdapterSpec() {
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: Should AdapterContext have this method?

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java (hunk 9 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java]: If `hasPlugin` return true, meaning the plugin is already added in this configurer from some prior call to one of those `usePlugin*` method. This is to avoid adding the same plugin with different properties (and potentially have the selector to select a different plugin).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java (hunk 10 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java]: AdapterSpecification is not in cdap-api yet. Hence no for now. We can move it there when we move AdapterSpecification.

File: cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java (hunk 7 above)
Comment (gokulavasan) [inline cdap-api/src/main/java/co/cask/cdap/api/templates/AdapterConfigurer.java]: Oh got it. ok.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: sounds good!

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java (hunk 4 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java]: oh ok.

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java (hunk 8 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/AdapterPlugin.java]: ok

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java (hunk 9 above)
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java]: So we can't we have two Transform plugins one after another? One say does filtering and another does say renaming?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java [hunk 11]
```diff
@@ -164,6 +169,50 @@ public TwillController launch(TwillApplication twillApplication) {
...
+
+    // Get all unique PluginInfo from the adapter spec
+    Set<PluginInfo> plugins = adapterSpec.getPluginInfos();
+
+    // If there is no plugin used by the adapter, nothing need to be localized
+    if (plugins.isEmpty()) {
+      return localizeFiles;
+    }
+
+    File templateDir = new File(cConf.get(Constants.AppFabric.APP_TEMPLATE_DIR));
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: so if original app.template.dir was set to /opt/cdap/master/plugins, templateDir will be plugins and pluginDir will be plugins/etlbatch ?

**Comment by gokulavasan** [comment]
Couple of questions. Primarily - https://github.com/caskdata/cdap/pull/2191#discussion_r28835717
Rest LGTM :+1:

File: cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java (hunk 9 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/templates/DefaultAdapterConfigurer.java]: hum.. so they are the same plugin but configure differently? hum.. probably need that.. Meaning when calling this method, you need to provide a unique name so that it can be used to associate with the property. Let me think about it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java (hunk 11 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: The `PluginInstantiator` works with this dir structure, hence maintaining the dir structure make the PluginInstantiator much simpler to be used in different mode.

**Comment by gokulavasan** [comment]
LGTM :+1:

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicMapReduceContext.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/DynamicMapReduceContext.java]: Do it in separate PR. In our current code base, usage of Nullable is not uniform.

**Comment by chtyim** [comment]
Addressed comments, fixed unit-test and rebased on develop.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT1576-3

**Comment by chtyim** [comment]
Import statement conflict resolved against develop. Squashed commits into one.
